File: src/channel.mjs
// Kanalmodi.
//
// Der wichtigste Unterschied zur ersten Konzeptfassung: LocalLink liefert
// nicht mehr nur "zuverlaessig und der Reihe nach". Ein reaktionsintensives
// Spiel, das seinen Zustand 30-mal pro Sekunde schickt, wird von genau dieser
// Zusage ruiniert - geht ein Paket verloren, wartet alles Nachfolgende auf die
// Wiederholung, obwohl es laengst ueberholt ist. Das ist Head-of-Line-Blocking,
// und es ist der Grund, warum Echtzeitspiele kein TCP sprechen.
//
// Deshalb waehlt die App den Modus pro Kanal.

/** Feuern und vergessen. Verlust ist egal, Reihenfolge ist egal. */
export const UNRELIABLE = 0;

/** Neuestes gewinnt. Veraltete Nachrichten werden verworfen statt zugestellt. */
export const SEQUENCED = 1;

/** Kommt an, Reihenfolge egal. Fuer unabhaengige Einzelereignisse. */
export const RELIABLE_UNORDERED = 2;

/** Kommt an, in Reihenfolge. Fuer Spielzuege, Lobby, Handshake. */
export const RELIABLE_ORDERED = 3;

export const MODE_NAMES = {
  [UNRELIABLE]: 'UNRELIABLE',
  [SEQUENCED]: 'SEQUENCED',
  [RELIABLE_UNORDERED]: 'RELIABLE_UNORDERED',
  [RELIABLE_ORDERED]: 'RELIABLE_ORDERED',
};

export function isReliable(mode) {
  return mode === RELIABLE_UNORDERED || mode === RELIABLE_ORDERED;
}

/**
 * Empfohlener Zuschnitt fuer ein Spiel. Kanal 0 bis 3 sind reserviert, damit
 * jede App auf denselben Steuerkanaelen redet und die Bibliothek sie kennt.
 */
export const SYSTEM_CHANNELS = {
  CONTROL: 0,   // Handshake, Faehigkeiten, Sitzungsverwaltung  RELIABLE_ORDERED
  HEARTBEAT: 1, // Lebenszeichen und Uhrenabgleich              UNRELIABLE
  EVENTS: 2,    // Spielereignisse, die ankommen muessen        RELIABLE_ORDERED
  STATE: 3,     // Zustandsschnappschuesse, neuestes gewinnt     SEQUENCED
};

export const DEFAULT_CHANNELS = [
  { id: SYSTEM_CHANNELS.CONTROL, mode: RELIABLE_ORDERED, coalesce: false },
  { id: SYSTEM_CHANNELS.HEARTBEAT, mode: UNRELIABLE, coalesce: true },
  { id: SYSTEM_CHANNELS.EVENTS, mode: RELIABLE_ORDERED, coalesce: false },
  { id: SYSTEM_CHANNELS.STATE, mode: SEQUENCED, coalesce: true },
];

File: src/commonMain/kotlin/dev/salvo/locallink/Channel.kt
package dev.salvo.locallink

/**
 * Kanalmodi. Die App waehlt pro Kanal, welche Zusage sie braucht - genau das
 * verhindert, dass ein reaktionsintensives Spiel an der Zuverlaessigkeit eines
 * einzelnen verlorenen Pakets haengen bleibt (Head-of-Line-Blocking).
 */
object ChannelMode {
    /** Feuern und vergessen. Verlust egal, Reihenfolge egal. */
    const val UNRELIABLE = 0

    /** Neuestes gewinnt. Veraltete Nachrichten werden verworfen. */
    const val SEQUENCED = 1

    /** Kommt an, Reihenfolge egal. */
    const val RELIABLE_UNORDERED = 2

    /** Kommt an, in Reihenfolge. */
    const val RELIABLE_ORDERED = 3

    fun isReliable(mode: Int): Boolean =
        mode == RELIABLE_UNORDERED || mode == RELIABLE_ORDERED

    fun name(mode: Int): String = when (mode) {
        UNRELIABLE -> "UNRELIABLE"
        SEQUENCED -> "SEQUENCED"
        RELIABLE_UNORDERED -> "RELIABLE_UNORDERED"
        RELIABLE_ORDERED -> "RELIABLE_ORDERED"
        else -> "UNKNOWN($mode)"
    }
}

/** Konfiguration eines Kanals. */
class ChannelConfig(
    val id: Int,
    val mode: Int,
    val coalesce: Boolean = false,
)

/**
 * Reservierte Steuerkanaele, damit jede App auf denselben Kanaelen redet und
 * die Bibliothek sie kennt.
 */
object SystemChannels {
    const val CONTROL = 0    // Handshake, Faehigkeiten       RELIABLE_ORDERED
    const val HEARTBEAT = 1  // Lebenszeichen, Uhrenabgleich  UNRELIABLE
    const val EVENTS = 2     // Spielereignisse               RELIABLE_ORDERED
    const val STATE = 3      // Zustandsschnappschuesse        SEQUENCED
}

val DEFAULT_CHANNELS: List<ChannelConfig> = listOf(
    ChannelConfig(SystemChannels.CONTROL, ChannelMode.RELIABLE_ORDERED, coalesce = false),
    ChannelConfig(SystemChannels.HEARTBEAT, ChannelMode.UNRELIABLE, coalesce = true),
    ChannelConfig(SystemChannels.EVENTS, ChannelMode.RELIABLE_ORDERED, coalesce = false),
    ChannelConfig(SystemChannels.STATE, ChannelMode.SEQUENCED, coalesce = true),
)
